akka-http-core/src/main/scala/akka/http/impl/engine/parsing/HttpMessageParser.scala

```diff
@@ -74,25 +78,31 @@ private[http] trait HttpMessageParser[Output >: MessageOutput <: ParserOutput] {
         case x             ⇒ x
       }
 
-    if (result.nonEmpty) throw new IllegalStateException("Unexpected `onPush`")
+    if (result ne null) throw new IllegalStateException("Unexpected `onPush`")
     run(state)
     doPull()
   }
 
   protected final def doPull(): Output =
-    if (result.nonEmpty) {
-      val head = result.head
-      result.remove(0) // faster than `ListBuffer::drop`
-      head
-    } else if (terminated) StreamEnd else NeedMoreData
+    result match {
+      case null ⇒ if (terminated) StreamEnd else NeedMoreData
+      case buffer: ListBuffer[Output] ⇒
+        val head = buffer.head
```

Comment:
What about this head vs indexed access at 0? Looks like the default head allocates two lambdas etc. and couldn't figure out if there was a specialized version somewhere.

Reply:
The `ListBuffer` variant is already the slow path so I didn't care about optimizing that as well.